### DISCUSSION TITLE
igl::copyleft::cgal::mesh_boolean -> really slow on tiny sample ... is it normal ?

### DISCUSSION THREAD

**Discussion by Vilo176** [libigl/libigl#1897] [Q&A]
Hi dear friends,

I'm testing LibIGL CSG capabilities on a bunch of test cases, and I'm facing very disappointing performances.
First of all, I'm compiling the last LibIGL source under Windows + VS2022, latest CGAL / Eigen / Boost.
All my tests are made in Release x64.

**Test 1 on sample to validate my compilation :**
cheburashka.off
decimated-knight.off
MESH_BOOLEAN_TYPE_RESOLVE
Triangles count = 23334
=> result is OK
=> computation time including IO = 2.5 seconds

**Test 2 on my own sample :**
input.obj
MESH_BOOLEAN_TYPE_RESOLVE
Triangles count = 292
=> result in not OK
=> computation time including IO = 85 seconds

Here are the operands.
All solids are intersecting each others, but are not self-intersecting.
![input](https://user-images.githubusercontent.com/25245090/134510008-69dde426-7dbf-4fb6-aa44-aaae7593f652.jpg)

Here is the result.
You can see that there are missing triangles, and that some triangles have not been splitted.
![output](https://user-images.githubusercontent.com/25245090/134510219-f92e99f5-5156-4164-ae28-afe98e3c520c.jpg)

And the OBJ
[input.zip](https://github.com/libigl/libigl/files/7217724/input.zip)

Maybe am I doing something wrong ...
Could you tell me if these result are "as expected" ?
Thanks :-)

**Comment by alecjacobson** [comment]
No, that doesn't sound right. When I try your `input.obj` it takes around 0.5 secs.

Are you really sure that you've compiled in Release mode? For Eigen and CGAL, it's very important to have the equivalent of `-O2`/`-O3` and `-DNDEBUG` flags.

**Comment by Vilo176** [comment]
Hi Alec,

Ok, after some digging I've found that my very poor performances was due to missing "tbb.dll" (for CGAL parallelism).
Overall computation time is now 0.5s on my sample.

Concerning anomalies in the output, do you have any advice ?

![bad1](https://user-images.githubusercontent.com/25245090/134646385-ac7c4ed3-4123-4474-b9ec-f78f5cfb3495.jpg)
![bad2](https://user-images.githubusercontent.com/25245090/134646390-c60acde3-fa24-4239-986c-1b5dfb220a1e.jpg)

**Reply by alecjacobson** [reply]
hmm, that's odd to me. I didn't think we were triggering any internal CGAL parallelism and our parallelism does not use TBB.

Meanwhile, is this the result of self-**_union_**  of `input.obj` ? If so, it looks incorrect.

(I noticed something funny for the **_resolve_** option called on this shape. I think this option needs to be revisited or removed. If that's the operation you're interested in, then using `remesh_self_intersections` directly would give the expected result in this case)

**Comment by Vilo176** [comment]
My final goal is to have a Mesh with all triangles intersections "resolved", I don't want to do CSG with it (for now).
I've redone the sample with remesh_self_intersections, result is still abnormal.

This is the same overlapping zone as above, you can see there are overlapping triangles :

![image](https://user-images.githubusercontent.com/25245090/134709233-e55651a9-e1c1-4eb4-abb5-0aad6ca1b964.png)

**Comment by alecjacobson** [comment]
Could you upload your output mesh from remesh_self_intersections? Ideally
as a double precision.ply rather than an .obj?
-Alec

**Comment by Vilo176** [comment]
Yes Sir,
I have simplified the object while preserving the "bad" behavior.

[sample.zip](https://github.com/libigl/libigl/files/7229489/sample.zip)

**Comment by Vilo176** [comment]
Concerning TBB I was wrong, as you expected, my slow computation times has nothing to do with it (module is not loaded during application lifetime).
I don't know what caused this behavior, but on my last release speed is normal (0.2s to compute above sample).

For completeness, here is my test code :

`void LIBIGL_Test(double* vertices, int vertices_nbr, int* triangles, int triangles_nbr)
{
	auto VV = Eigen::MatrixXd(vertices_nbr/3, 3);
	for (int i=0; i<vertices_nbr/3; i++)
		for (int j=0; j<3; j++)
			VV(i,j) = vertices[3*i+j];

	auto FF = Eigen::MatrixXi(triangles_nbr/3, 3);
	for (int i=0; i<triangles_nbr/3; i++)
		for (int j=0; j<3; j++)
			FF(i,j) = triangles[3*i+j];

	igl::writePLY("input.ply", VV, FF);

	Eigen::MatrixXd Vr;
	Eigen::MatrixXi Fr;
	Eigen::MatrixXi IF;
	Eigen::VectorXi CJ;
	Eigen::VectorXi I;
	igl::copyleft::cgal::remesh_self_intersections(VV, FF, igl::copyleft::cgal::RemeshSelfIntersectionsParam(false, false, true), Vr, Fr, IF, CJ, I);

	igl::writePLY("output.ply", Vr, Fr);
}`

**Reply by Vilo176** [reply]
Hi Alec, any news or hint to solve my issue ?

**Reply by alecjacobson** [reply]
can you upload the double precision .ply?

**Reply by Vilo176** [reply]
In my post above I've uploaded "sample.zip" containing both "input.ply" and "output.ply" as produced by Libigl writer.
Maybe you don't have access to it ?

=> https://github.com/libigl/libigl/files/7229489/sample.zip

**Comment by Vilo176** [comment]
Hi Alec,

Could you please confirm that you have access to the tiny sample that shows my "issue" ?
I guess you are very busy, but at least I want to be sure that you have all infos to, at some moment, take a look at it.
bye

**Comment by Nosenzor** [comment]
I just come to this thread and I opened your meshes. @Vilo176 did you notice your input was not a (set of ) closed meshes ? I'm not sure it is the problem here but I found that can be interesting to share 
<img width="959" alt="image" src="https://user-images.githubusercontent.com/2452388/203380666-99e8f148-b99b-40a2-9600-34787fba381e.png">

**Comment by Vilo176** [comment]
As far as I can tell, MESH_BOOLEAN_TYPE_RESOLVE is all about splitting triangles.
Thus, no matter if triangles are actually representing closed manifold solids or not.
What I was expecting is just LibIGL to compute split parts of all triangles.

By the way, thanks for your investigations !

**Reply by cljbb** [reply]
I encountered the same problem. remesh_self_intersections is very slow in my Windows program. The same program was complied on Mac and Linux. They are very fast. Only the Windows version has the problem.

Have you found any cause of this and solutions.

**Reply by Vilo176** [reply]
I'm afraid I have not dug further into libigl ...

**Reply by cljbb** [reply]
Thank you for letting me know. Did you use other libraries for solving mesh intersections and mesh boolean?

I am just curious. I did some search and found that libigl's interface is really easy to use; however, it is just that performance problem.

**Reply by Vilo176** [reply]
No I don't.
All this stuff is for research, not for real life cases.
It cannot be included into production grade products (and this is perfectly fair).
Good luck ;-)

**Reply by cljbb** [reply]
Thank you for the information, Vilo.
